src/yunohost/migrations.py

```diff
@@ -0,0 +1,249 @@
+
+    migrations = []
+
+    # loading all migrations
+    for migration in migrations_list()["migrations"]:
+        logger.debug("Loading migration {number} {name}...".format(
+            number=migration["number"],
+            name=migration["name"],
+        ))
+
+        try:
+            module = import_module("yunohost.data_migrations.{file_name}".format(**migration))
+        except Exception:
+            import traceback
+            traceback.print_exc()
+
+            raise MoulinetteError(errno.EINVAL, "WARNING: failed to load migration {number} {name}".format(
+                number=migration["number"],
+                name=migration["name"],
+            ))
+            break
+
+        migrations.append({
+            "number": migration["number"],
+            "name": migration["name"],
+            # this is python builtin method to import a module using a name, we use that to import the migration
+            # has a python object so we'll be able to run it in the next loop
+            "module": module,
+        })
+
+    migrations = sorted(migrations, key=lambda x: x["number"])
+
+    if not migrations:
+        logger.info("No migrations to run.")
+        return
+
+    all_migration_numbers = [x["number"] for x in migrations]
+
+    if target is None:
+        target = migrations[-1]["number"]
+
+    # validate input, target must be "0" or a valid number
+    elif target != 0 and target not in all_migration_numbers:
+        raise MoulinetteError(errno.EINVAL, "Invalide number for target argument, available migrations numbers are 0 or {}".format(", ".join(map(str, all_migration_numbers))))
+
+    logger.debug("migration target is {}".format(target))
+
+    # no new migrations to run
+    if target == last_runed_migration_number:
+        logger.warn("no migrations to run")
+        return
+
+    logger.debug("last runed migrations is {}".format(last_runed_migration_number))
+
+    # we need to run missing migrations
+    if last_runed_migration_number < target:
+        logger.debug("migrating forward")
+        # drop all already runed migrations
+        migrations = filter(lambda x: x["number"] > last_runed_migration_number, migrations)
+        mode = "forward"
+
+    # we need to go backward on already runed migrations
+    elif last_runed_migration_number > target:
+        logger.debug("migrating backward.")
+        # drop all not already runed migrations
+        migrations = filter(lambda x: x["number"] <= last_runed_migration_number, migrations)
+        mode = "backward"
+
+    else:  # can't happend, this case is handle before
+        raise Exception()
+
+    # effectively run selected migrations
+    for migration in migrations:
+        if not fake:
+            logger.warn("Runing migration {number} {name}...".format(**migration))
+
+            try:
+                if mode == "forward":
+                    migration["module"].MyMigration().migrate()
```

Comment:
Really a minor detail / suggestion, but maybe it could be more beautiful to call just `migration["module"].MyMigration().migrate(mode)`, then in the class, `if mode == "forward": self.forward()` and so on.